django/db/models/base.py

```diff
@@ -869,9 +869,10 @@ def _save_table(self, raw=False, cls=None, force_insert=False,
                 # autopopulate the _order field
                 field = meta.order_with_respect_to
                 filter_args = field.get_filter_kwargs_for_object(self)
-                order_value = cls._base_manager.using(using).filter(**filter_args).count()
-                self._order = order_value
-
+                max_order_value = cls._base_manager.using(using).filter(
+                    **filter_args
+                ).aggregate(Max('_order'))['_order__max']
+                self._order = max_order_value + 1 if max_order_value is not None else 0
```

Comment:
We could probably use more complicated expression to get rid of a temporary variable to make it more "atomic", e.g.
```python
self._order = cls._base_manager.using(using).filter(**filter_args).aggregate(
    max_order=Coalesce(
        ExpressionWrapper(Max('_order') + Value(1), output_field=IntegerField()),
        Value(0),
    ),
)['max_order']
```
:thinking: